c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -3702,6 +3704,382 @@ bool init_class(JSContext *cx, HandleObject global) {
+  return output;
+}
+
+inline JS::Result<std::string> base64Decode2to1(std::string input) {
+  std::string output = "";
+  uint8_t w, x;
+  // 8.1 Find the code point pointed to by position in the second column of Table 1: The Base 64
+  // Alphabet of RFC 4648. Let n be the number given in the first cell of the same row. [RFC4648]
+  if (!base64CharacterToValue(input[0], &w) || !base64CharacterToValue(input[1], &x)) {
+    return JS::Result<std::string>(JS::Error());
+  }
+  // 9. If buffer is not empty, it contains either 12 or 18 bits. If it contains 12 bits, then
+  // discard the last four and interpret the remaining eight as an 8-bit big-endian number. If it
+  // contains 18 bits, then discard the last two and interpret the remaining 16 as two 8-bit
+  // big-endian numbers. Append the one or two bytes with values equal to those one or two numbers
+  // to output, in the same order.
+  output += (uint8_t(w << 2 | x >> 4));
+  return output;
+}
+// https://infra.spec.whatwg.org/#forgiving-base64-decode
+JS::Result<std::string> forgivingBase64Decode(std::string input) {
+  // 1. Remove all ASCII whitespace from data.
+  // ASCII whitespace is U+0009 TAB, U+000A LF, U+000C FF, U+000D CR, or U+0020 SPACE.
+  std::regex whitespace("[\t\n\f\r ]+");
+  input = std::regex_replace(input, whitespace, "");
+  uint32_t inputLength = input.length();
+
+  // 2. If data’s code point length divides by 4 leaving no remainder, then:
+  if (inputLength && (inputLength % 4 == 0)) {
+    // 2.1 If data ends with one or two U+003D (=) code points, then remove them from data.
+    if (input[inputLength - 1] == '=') {
+      if (input[inputLength - 2] == '=') {
+        inputLength -= 2;
+      } else {
+        inputLength -= 1;
+      }
+    }
+  }
+
+  // 3. If data’s code point length divides by 4 leaving a remainder of 1, then return failure.
+  if ((inputLength % 4 == 1)) {
+    return JS::Result<std::string>(JS::Error());
+  }
+
+  // 4. If data contains a code point that is not one of
+  //    U+002B (+)
+  //    U+002F (/)
+  //    ASCII alphanumeric
+  // then return failure.
+
+  // Step 4 is handled within the calls below to
+  // base64Decode4to3, base64Decode3to2, and base64Decode2to1
+
+  // 5. Let output be an empty byte sequence.
+  std::string output = "";
+
+  // 6. Let buffer be an empty buffer that can have bits appended to it.
+
+  // Step 6 is handled within the calls below to
+  // base64Decode4to3, base64Decode3to2, and base64Decode2to1
+
+  // 7. Let position be a position variable for data, initially pointing at the start of data.
+
+  // We don't use a position variable, instead we erase from the `input` each time we have dealt
+  // with some characters.
+
+  while (inputLength >= 4) {
+    auto out_result = base64Decode4to3(input);
+    if (out_result.isErr()) {
+      return JS::Result<std::string>(JS::Error());
+    }
+    output += out_result.unwrap();
+
+    input.erase(0, 4);
+    inputLength -= 4;
+  }
```

Comment:
What do you think about restructuring this loop to operate on a `std::string_view` instead? This way you won't incur any cost for dropping data off of the front of the buffer like you might with `erase`, and you can ensure that none of the decoding functions will modify your buffer. You'll also avoid copying the input buffer when calling the decoding functions, as `std::string_view` is very cheap to copy around.
```suggestion
  std::string_view buffer = input;                                                   
  while (buffer.size() >= 4) {                                                       
    auto out_result = base64Decode4to3(input.substr(0, 4));                          
    if (out_result.isErr()) {                                                        
      return JS::Result<std::string>(JS::Error());                                   
    }                                                                                
    output += out_result.unwrap();                                                   
                                                                                     
    buffer = buffer.substr(5);                                                       
  } 
```

Reply:
Great suggestion - this makes complete sense, string_view is the perfect type for this ☺️ 
I made this change in a single commit alongside updating all the decode functions to use string_view - [`54c2eec` (#227)](https://github.com/fastly/js-compute-runtime/pull/227/commits/54c2eecdd20cfd2d1e2fb43b8efbfb2552f6fe05)

I'll do the same change for the encoding functions as well 👍